chalice/chalicelib/matrix_handler.py

```diff
@@ -0,0 +1,165 @@
+        """
+        Merge matrices within bundles, and upload the merged matrix to an s3 bucket.
+
+        :param bundle_uuids: Bundles' uuid for locating bundles in DSS.
+        :param request_id: Merge request id.
+        :param job_id: Job id of the request.
+        """
+        try:
+            logger.info("Concatenate matrices for request_id: {}".format(request_id))
+
+            # Print out the current usage of /tmp/ directory
+            call(["df", "-H", TEMP_DIR])
+
+            logger.info("tmp directory contains: {}".format(str(os.listdir(TEMP_DIR))))
+
+            # Clean /tmp folder before each run
+            call('rm -rf /tmp/*', shell=True)
+
+            logger.info("After cleaning, /tmp directory contains: {}".format(str(os.listdir(TEMP_DIR))))
+
+            # Update the request status to RUNNING
+            RequestHandler.update_request(
+                bundle_uuids=bundle_uuids,
+                request_id=request_id,
+                job_id=job_id,
+                status=RequestStatus.RUNNING
+            )
+
+            # Create a temp directory for storing all temp files
+            with tempfile.TemporaryDirectory(dir=TEMP_DIR, prefix="{}_".format(request_id)) as temp_dir:
+
+                start_time = time.time()
+                mtx_paths = self._download_mtx(bundle_uuids=bundle_uuids, temp_dir=temp_dir)
+                _, merged_mtx_path = tempfile.mkstemp(dir=temp_dir, prefix=request_id, suffix=self._suffix)
+                self._concat_mtx(mtx_paths=mtx_paths, out_file=merged_mtx_path)
+                self._upload_mtx(path=merged_mtx_path, request_id=request_id)
+                end_time = time.time()
+
+                # Update the request status to DONE
+                RequestHandler.update_request(
+                    bundle_uuids=bundle_uuids,
+                    request_id=request_id,
+                    job_id=job_id,
+                    status=RequestStatus.DONE,
+                    time_spent_to_complete="{} seconds".format(end_time - start_time)
+                )
+
+        except Exception as e:
+            temp_files = os.listdir(TEMP_DIR)
+            logger.info("tmp directory contains: {}".format(str(temp_files)))
+
+            # Update the request status to ABORT
+            RequestHandler.update_request(
+                bundle_uuids=bundle_uuids,
+                request_id=request_id,
+                job_id=job_id,
+                status=RequestStatus.ABORT,
+                reason_to_abort=traceback.format_exc()
+            )
+
+            raise e
+
+
+class LoomMatrixHandler(MatrixHandler):
+    """
+    Matrix handler for .loom file format
+    """
+
+    def __init__(self) -> None:
+        super().__init__(".loom")
+
+    def _concat_mtx(self, mtx_paths: List[str], out_file: str) -> None:
+        try:
+            logger.info("Combining matrices to %s.", out_file)
+            loompy.combine(mtx_paths, out_file)
```

Comment:
Looks like there's a `batch_size` argument to this that we could try cranking up.

Reply:
Okay. Do you have any recommended value that I can use for `batch_size`? Or do you think I can conduct some experiments on it to find the most optimal one?

Reply:
Probably need to experiment a little. I just recall that you were a little surprised at the speed of the merges, so it might be worth trying to increase it. I can say anecdotally that increasing the size/reducing the number of reads and writes can improve HDF5 I/O a lot.

Reply:
Sure. I will do that.